core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1272,8 +1272,8 @@ protected Engine engineUnsafe() {
         @Override
         public void onFailedEngine(ShardId shardId, String reason, @Nullable Throwable failure) {
             try {
-                // delete the shard state so this folder will not be reused
-                MetaDataStateFormat.deleteMetaState(nodeEnv.availableShardPaths(shardId));
+                // mark as corrupted, so opening the store will fail
+                store.markStoreCorrupted(new IOException(reason, failure));
```

Comment:
Can we add some text so we know it came from here regardless of the reason? Something like "failed engine (reason ["+reason +"])"?